bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+
+    // Handle a `RequestType::Learn` from a peer
+    fn on_learn(&mut self, now: Instant, from: Baseboard, request_id: Uuid) {
+        let err = match &self.state {
+            State::Uninitialized => Some(MsgError::NotInitialized),
+            State::Learning => Some(MsgError::StillLearning),
+            State::Learned { .. } => Some(MsgError::CannotSpareAShare),
+            State::InitialMember { pkg, .. } => {
+                self.request_manager.new_learn_received_req(
+                    request_id,
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    from.clone(),
+                    &self.connected_peers,
+                );
+                // We need to gather shares and reconstruct the rack secret.
+                // Therefore we don't send a response immediately.
+                None
+            }
+        };
+        if let Some(err) = err {
+            self.push_response(from, request_id, err.into());
+        }
+    }
+
+    // Handle a `Response` from a peer
+    fn handle_response(
+        &mut self,
+        from: Baseboard,
+        rsp: Response,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match rsp.type_ {
+            ResponseType::InitAck => self.on_init_ack(from, rsp.request_id),
+            ResponseType::Share(share) => {
+                self.on_share(from, rsp.request_id, share)
+            }
+            ResponseType::Pkg(pkg) => self.on_pkg(from, rsp.request_id, pkg),
+            ResponseType::Error(error) => {
+                Err(ApiError::ErrorResponseReceived {
+                    from,
+                    state: self.state.name(),
+                    request_id: rsp.request_id,
+                    error,
+                })
+            }
+        }
+    }
+
+    // Handle a `ResposneType::InitAck` from a peer
+    fn on_init_ack(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match self.request_manager.on_init_ack(from.clone(), request_id) {
+            Some(true) => Ok(Some(ApiOutput::RackInitComplete)),
+            Some(false) => Ok(None),
+            None => Err(ApiError::UnexpectedResponse {
+                from,
+                state: self.state.name(),
+                request_id,
+                msg: "InitAck",
+            }),
+        }
+    }
+
+    // Handle a `ResponseType::Pkg` from a peer
+    fn on_pkg(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        pkg: LearnedSharePkg,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        if self.request_manager.on_pkg(request_id) {
+            // This pkg matched our outstanding request. Let's transition from
+            // `State::Learning` to `State::Learned`.
+            assert_eq!(self.state, State::Learning);
+            self.state = State::Learned { pkg };
+            Ok(Some(ApiOutput::LearningCompleted))
+        } else {
```

Comment:
Nit: you could flatten this into `} else if self.state == State::Learning {`

(If you run Clippy, it will probably complain about this and other things too)

Reply:
done, and clippied